### PR TITLE
Use precomputed values for FB expansion of filter objects

### PR DESCRIPTION
**Pull request by junchaoxia**
This fixes the issue #199 using precomputed values for the FFB expansion of filter objects.

### CODE REVIEW

File: src/aspire/basis/ffb_2d.py [hunk 1]
```diff
@@ -65,13 +66,24 @@ def _precomp(self):
...
             for k in range(1, self.k_max[ell] + 1):
-                radial[:, ind_radial] = jv(ell, self.r0[k - 1, ell] * r / self.kcut)
+                jv_ell_k = jv(ell, self.r0[k - 1, ell] * r / self.kcut)
                 # NOTE: We need to remove the factor due to the discretization here
                 # since it is already included in our quadrature weights
                 nrm = 1 / (np.sqrt(np.prod(self.sz))) * self.basis_norm_2d(ell, k)
-                radial[:, ind_radial] /= nrm
+                jv_ell_k_norm = jv_ell_k / nrm
+                radial[:, ind_radial] = jv_ell_k_norm
+                # CTF filters use different normalized factors
```
Comment (janden) [inline src/aspire/basis/ffb_2d.py]: I don't quite understand. Why should the normalization be different here? Are you sure this is not just a bug in the `filter_to_fb_mat` code?

**Review by janden — CHANGES_REQUESTED** [review]

File: src/aspire/basis/ffb_2d.py (hunk 1 above)
Comment (junchaoxia) [reply src/aspire/basis/ffb_2d.py]: There filer_to_fb_mat is almost line to line converting from the original Matlab as below, 
  https://github.com/PrincetonUniversity/aspire/blob/master/projections/radial_filter_to_fb_mat.m
I compared the results and they are consistent when I did the Cov2D.

Comment (janden) [reply src/aspire/basis/ffb_2d.py]: Well there might have been a bug in the Matlab code as well. The question remains, why should the normalization be different? Do we get the wrong result when we use the same normalization?

Comment (janden) [reply src/aspire/basis/ffb_2d.py]: Let's make a separate issue for the normalization and have this PR focus on optimizing the code.

Since the required values already exist in `basis.radial`, let's use those in `filter_to_fb_mat` by multiplying with the right constant to get the desired normalization factor. There's no need to store them in a separate `jv_norm` attribute.

Comment (janden) [reply src/aspire/basis/ffb_2d.py]: Also, here we're relying on the fact that we use the same number of radial points as in `basis._precomp`. That will not necessarily be the case always. To make this dependence explicit, we either need to store that value `n_r` as an attribute of `basis` that can then be retrieved by `filter_to_fb_mat` (with a line `n_k = fbasis.n_r`) or just rely on the fact that `n_k = fbasis.r0.shape[0]`, but I find this a less clear.

Comment (junchaoxia) [reply src/aspire/basis/ffb_2d.py]: I checked again. the Matlab version has the implementations as the Python version. In the paper of FFB basis, the normalization factor is `N_lk = 1/(c*sqrt(pi) *J_(l+1)(R_lk))`,  with c=0.5 for radius cutoff, and J for the Bessel function. This normalization factor used for filter expanding, `N_lk = 1/(c *J_(l+1)(R_lk)*sqrt(1/2)`, does not include the sqrt(pi) but addition sqrt(1/2) as shown below: 
https://github.com/PrincetonUniversity/aspire/blob/f152e8efd9856f75c294f453ef7846bd3a684410/projections/radial_filter_to_fb_mat.m#L43

In the implementation of FFB basis,  N_lk = 1/(c*sqrt(pi) *J_(l+1)(R_lk)*sqrt(1/2)), with additional sqrt(1/2) to the paper as shown
https://github.com/PrincetonUniversity/aspire/blob/master/basis/ffb_basis.m#L275
https://github.com/PrincetonUniversity/aspire/blob/master/basis/ffb_basis.m#L182

The outputs from the Cob2d_simulation example with current implementation of different normalization factors are list as below:

> Deviation of the noisy mean estimate: 0.008807039379139734
> 2020/08/20 15:41:03 Deviation of the noisy covariance estimate: 0.032321671019596605
> 2020/08/20 15:41:03 Estimated images normalized RMSE: 0.22725983789333037

Using the same normalization factor from Filter expansion, the outputs are 

> Deviation of the noisy mean estimate: 0.0088070393791396
> 2020/08/28 22:11:10 Deviation of the noisy covariance estimate: 0.22049860794751824
> 2020/08/28` 22:11:10 Estimated images normalized RMSE: 3.713621537096643

Using the same normalization factor from FFB basis, the outputs are 

> 2020/08/28 23:10:40 Deviation of the noisy mean estimate: 5.267011983601262
> 2020/08/28 23:10:40 Deviation of the noisy covariance estimate: 16.983344615954522
> 2020/08/28 23:10:40 Estimated images normalized RMSE: 3.399225365537338

In summary, the current implementation with different normalization factors have lower errors.  

Regarding performance improvements, the `filter_to_fb_mat function` used to take 5.8%(1426ms) wall time of the cov2d_simulation example on my 8 core laptop. Now it reduces to 0.2%(66ms) using precomputed values for Bessel functions on my 8 core laptop.

Comment (junchaoxia) [reply src/aspire/basis/ffb_2d.py]: Made `n_r` and `n_theta` as attributes in FFB2D basis class in new commit e7980ce. Currently the `fbasis._precomp['radial']` have the shape of `(n_r, np.sum(self.k_max))` which needs to be renormalized and reshaped to the embedded list of `[ell][k][n_r]` for the double loops in the `filter_to_fb_mat` function if we remove the `jv_norm` attribute. This includes multiplying different normalization factors for ell=0 and using two loops to reshape since each `ell` has different `k_max`. I think using `jv_norm` in current state is simpler. Please let me know if you prefer to remove `jv_norm` and rewriting the `filter_to_fb_mat` function.

Comment (junchaoxia) [reply src/aspire/basis/ffb_2d.py]: Creates new issue #250 for understanding the different normalization factors.

Comment (janden) [reply src/aspire/basis/ffb_2d.py]: There shouldn't be any need for reshaping, just indexing and renormalizing of `_precomp['radial']` here. Having the same data represented in two different ways is not simpler. Please get rid of `jv_norm` and use `_precomp['radial']` directly.

Comment (junchaoxia) [reply src/aspire/basis/ffb_2d.py]: `jv_norm` was removed in new commit  190914d.

File: src/aspire/utils/blk_diag_matrix.py [hunk 2]
```diff
@@ -855,22 +854,28 @@ def filter_to_fb_mat(h_fun, fbasis):
...
     for ell in range(0, fbasis.ell_max+1):
         k_max = fbasis.k_max[ell]
         rmat = 2*k_vals.reshape(n_k, 1)*fbasis.r0[0:k_max, ell].T
         fb_vals = np.zeros_like(rmat)
         for ik in range(0, k_max):
-            fb_vals[:, ik] = jv(ell, rmat[:, ik])
-        fb_nrms = 1/np.sqrt(2)*abs(jv(ell+1, fbasis.r0[0:k_max, ell].T))/2
-        fb_vals = fb_vals/fb_nrms
+            # filters use different normalized factors
+            if ell == 0:
```
Comment (janden) [inline src/aspire/utils/blk_diag_matrix.py]: The only difference between the two branches is the factor `np.sqrt(2)`. Move the copy from `fbasis._precomp` out and have a single `if ell == 0` statement that multiplies by `np.sqrt(2)`.

**Review by janden — CHANGES_REQUESTED** [review]

File: src/aspire/utils/blk_diag_matrix.py (hunk 2 above)
Comment (junchaoxia) [reply src/aspire/utils/blk_diag_matrix.py]: Done. The inner loop was also removed in commit 01c8fcd.

**Review by janden — DISMISSED** [review]
Ok looks good. Please don't merge this until we discuss on Wednesday.

File: src/aspire/basis/ffb_2d.py [hunk 3]
```diff
@@ -59,23 +62,21 @@ def _precomp(self):
...
+        n_r = self.n_r
+        n_theta = self.n_theta
         r, w = lgwt(n_r, 0.0, self.kcut)
 
         radial = np.zeros(shape=(n_r, np.sum(self.k_max)))
         ind_radial = 0
         for ell in range(0, self.ell_max + 1):
             for k in range(1, self.k_max[ell] + 1):
-                radial[:, ind_radial] = jv(ell, self.r0[k - 1, ell] * r / self.kcut)
+                jv_ell_k = jv(ell, self.r0[k - 1, ell] * r / self.kcut)
```
Comment (janden) [inline src/aspire/basis/ffb_2d.py]: Please revert this and the next change. This doesn't add anything right now and will create more conflicts with #231.

Comment (junchaoxia) [reply src/aspire/basis/ffb_2d.py]: Done in commit cb4bec6.

**Review by janden — DISMISSED** [review]

File: src/aspire/utils/blk_diag_matrix.py [hunk 4]
```diff
@@ -860,7 +861,7 @@ def filter_to_fb_mat(h_fun, fbasis):
         rmat = 2*k_vals.reshape(n_k, 1)*fbasis.r0[0:k_max, ell].T
         fb_vals = np.zeros_like(rmat)
         ind_radial = np.sum(fbasis.k_max[0:ell])
-        fb_vals[:, 0:k_max] = fbasis._precomp['radial'][:,
+        fb_vals[:, 0:k_max] = radial[:,
```
Comment (janden) [inline src/aspire/utils/blk_diag_matrix.py]: Can we keep this on one line? If not, can we break it in a more natural spot?

**Review by janden — CHANGES_REQUESTED** [review]

File: src/aspire/utils/blk_diag_matrix.py (hunk 4 above)
Comment (junchaoxia) [reply src/aspire/utils/blk_diag_matrix.py]: I tried to limit the line to below 76 characters without using `\` for continuing. Changed to one line in new commit.

Comment (janden) [reply src/aspire/utils/blk_diag_matrix.py]: Yes but you can break before the `*` by enclosing in parentheses instead of breaking in the middle of the indices.

Comment (junchaoxia) [reply src/aspire/utils/blk_diag_matrix.py]: If breaking at  `*` after radial array,  Pycharm will automatically add `\` to make the next line as continuing one without ambiguity.  Anyway, only use one line in the new version. We might use some tool to limit all code to the style of 76 characters if we really need it in future.

**Review by janden — DISMISSED** [review]

**Comment by junchaoxia** [comment]
@janden The PR was rebased and updated based on the PR# 231. Please take a quick look.

**Review by janden — APPROVED** [review]
